### ISSUE TITLE
Defaulting to allowed_push_host doesn't quite work

### ISSUE THREAD

**Issue by randycoulman** [rubygems/rubygems#1787]
I'm having a problem.

My current problem is: When I attempt to push a private gem to a private gem server, I am still required to specify the `--host` argument, even though #1486 attempts to default the host to the `allowed_push_host` configured in the `gemspec`.

If I run:

```
gem push pkg/my-test-gem-0.1.0.gem
```

I see:

```
Enter your  credentials.
Don't have an account yet? Create one at /sign_up
   Email:   
```

According to #1486, it should default to using my configured `allowed_push_host`, but instead is prompting for my rubygems.org credentials.

I believe that the root cause of this is that at https://github.com/rubygems/rubygems/blob/0749715e4bd9e7f0fb631a352ddc649574da91c1/lib/rubygems/commands/push_command.rb#L47, the `push` command is calling `sign_in @host`, but `@host` is `nil` at this point.  It hasn't yet been defaulted to the `allowed_push_host`.

That defaulting isn't done until https://github.com/rubygems/rubygems/blob/0749715e4bd9e7f0fb631a352ddc649574da91c1/lib/rubygems/commands/push_command.rb#L76-L86 in the `send_gem` method that is called just after the `sign_in` call.

A possible fix would be to not make the `sign_in` call until after the defaulting is done, but I'm not sure if that would cause other issues.

Additional details:

In my `my-test-gem.gemspec` file, I have:

```
spec.metadata['allowed_push_host'] = "http://my-gemstash-server:9292/private"
```

In my `.gemrc` file, I have:

```
:disable_default_gem_server: true
```

And, in my `.gem/credentials` file, I have:

```
http://my-gemstash-server:9292/private: <my generated key>
```

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [X] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
$ gem env version
2.6.8
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by segiddins** [comment]
Best fix is probably to do `@host = options[:host] || gem_data.spec.metadata['default_gem_server'] || gem_data.spec.metadata['allowed_push_host'] ` or something like that?

**Comment by randycoulman** [comment]
@segiddins Are you thinking that code should go in `execute`?  If so, we'd have to change where/when we get a hold of `gem_data`, because it isn't available until here: https://github.com/rubygems/rubygems/blob/0749715e4bd9e7f0fb631a352ddc649574da91c1/lib/rubygems/commands/push_command.rb#L74

To get it earlier, we'd also need to call `get_one_gem_name` earlier as well.  I don't see any danger in that.

Looking more closely, we could move all of the use of `gem_data` up to `execute` and compute `@host` like you suggest.

But the part I don't understand yet is here: https://github.com/rubygems/rubygems/blob/0749715e4bd9e7f0fb631a352ddc649574da91c1/lib/rubygems/commands/push_command.rb#L89

We're passing both `@host` and `push_host` to the API request.  I'm not sure what that does or how important it is to preserve that behavior.

**Comment by segiddins** [comment]
@randycoulman yeah, that's what I'm guessing will work, but I haven't looked into the specifics too deeply

**Comment by bronzdoc** [comment]
@segiddins @randycoulman what you think about moving `sign_in @host` inside `send_gem` after @host is defined?

**Comment by randycoulman** [comment]
@bronzdoc I think that would work.  I experimented on a branch last week that moved the host/push_host computation up into `execute`, but haven't had time to finish it yet.  It seems to work when I run it, but it breaks some tests and I haven't had time to dig back in and figure that part out, nor to write a new failing test for this case.

**Comment by randycoulman** [comment]
I could use a hand with the test infrastructure here.  Most of the test for the push command bypass the `execute` method and test the `send_gem` method directly, so pulling the host defaulting code up to execute breaks a bunch of tests.

It seems like I have two options:

* Modify the tests to go through `execute` instead; I started by modifying the top part of `send_battery` to look like this:

```ruby
use_ui @ui do
  @cmd.options[:args] = [@path]
  @cmd.instance_variable_set :@host, @host
  @cmd.execute
end
```

but it failed a number of tests with errors that were hard to decipher (things like `FetchError: no data` and `TermError` and mismatched output.

* Move the `sign_in` call down to `send_gem`.  This surprisingly doesn't fail any tests.  I expected it to fail trying to sign into a host.

Either way, I'd like to modify the tests to verify that we try to sign in to the correct host.  Is there a recommended way in this codebase to verify the `sign_in` host?  I see the tests for `sign_in` itself, but that seems like a lot of setup for this.

Any suggestions?

**Comment by bronzdoc** [comment]
@randycoulman I haven't looked at this deeply yet and don't have experience with those `push command` tests to suggest something :( but I'll look at this issue tonight :).

**Comment by randycoulman** [comment]
Ping?  Can anyone help me with my test infrastructure question above?  I'd like to get my PR finished and submitted.  Thanks!

**Comment by bronzdoc** [comment]
@randycoulman sorry this has took so long... do you have a working branch i can peek to try to close this? thanks!

**Comment by randycoulman** [comment]
@bronzdoc Thanks for taking a look at this!

I found the work I'd been doing earlier.  I don't have time to update to the latest master right now, but I pushed up the branch here: https://github.com/randycoulman/rubygems/tree/default-to-allowed-push-host

I vaguely remember doing some additional experimenting with different approaches as I outlined above, but I don't seem to have kept any of the code for that.

If this is enough for you to go on, great.  If not, let me know and I'll try to set aside some time to take another run at this.  Probably won't be for a couple of weeks, though, as I have some other commitments and a short vacation coming up.
